lib/util/make-knex.js

```diff
@@ -140,83 +213,18 @@ function redefineProperties(knex, client) {
   // any other information is specified.
   QueryInterface.forEach(function(method) {
     knex[method] = function() {
-      const builder = knex.queryBuilder();
+      const builder = this.queryBuilder();
       return builder[method].apply(builder, arguments);
     };
   });
 
-  Object.defineProperties(knex, {
-    context: {
-      get() {
-        return knex._context;
-      },
-      set(context) {
-        knex._context = context;
-
-        // Redefine public API for knex instance that would be proxying methods from correct context
-        knex.raw = context.raw;
-        knex.batchInsert = context.batchInsert;
-        knex.transaction = context.transaction;
-        knex.transactionProvider = context.transactionProvider;
-        knex.initialize = context.initialize;
-        knex.destroy = context.destroy;
-        knex.ref = context.ref;
-        knex.withUserParams = context.withUserParams;
-        knex.queryBuilder = context.queryBuilder;
-        knex.disableProcessing = context.disableProcessing;
-        knex.enableProcessing = context.enableProcessing;
-      },
-      configurable: true,
-    },
-
-    client: {
-      get() {
-        return knex.context.client;
-      },
-      set(client) {
-        knex.context.client = client;
-      },
-      configurable: true,
-    },
-
-    userParams: {
-      get() {
-        return knex.context.userParams;
-      },
-      set(userParams) {
-        knex.context.userParams = userParams;
-      },
-      configurable: true,
-    },
-
-    schema: {
-      get() {
-        return knex.client.schemaBuilder();
-      },
-      configurable: true,
-    },
-
-    migrate: {
-      get() {
-        return new Migrator(knex);
-      },
-      configurable: true,
-    },
-
-    seed: {
-      get() {
-        return new Seeder(knex);
-      },
-      configurable: true,
-    },
+  for (const m of CONTEXT_METHODS) {
+    knex[m] = function(...args) {
+      return this.context[m](...args);
+    };
+  }
```

Comment:
This was really the key change.  Previously, these methods were not actually being proxied.  Instead, they would be injected directly into the `knex` function.  For example:

```
knex.transaction = context.transaction;
knex.raw = context.raw
```

The problem: this meant that the methods would be invoked through `knex` now instead of `context`.  Thus, `this` would point to `knex` instead of `context`.  As a result, it was impossible to define methods in `context` such as:

```
const context = {
  transaction(container, _options) {
    // Do some implementation-specific stuff
    const outerTx = createOuterTxForTopLevelTransactions();

    // This line would fail because `this` is pointing to `knex` instead of `context`!
    return this._transaction(container, options, outerTx);
  },

  _transaction(container, options, outerTx) {
    // Do stuff that is common across all implementations...
  }
}
```